### ISSUE TITLE
Chessboard issue with output layer(s)

### ISSUE THREAD

**Issue by aaparab** [aaparab/png2pgn#1]
I'm facing a problem with how to organize the last layer. For each of the 64 squares, I need to use a softmax to identify whether the piece in that square is among empty, white king, white queen, ..., black pawn (13 possibilities). The attempt for a vanilla neural network is outlined [here](../blob/master/code/golf.ipynb). You are welcome to discuss.

**Comment by sgsawant** [comment]
The link (see *outlined here* in OP) doesn't work for me.

On a related note, I would say it's better to write a small lightweight
function which can tell whether the square is empty or occupied. The
primary impact this will have is that more training muscle will be devoted
to distinguishing between the remaining classes; information stored within
the network will be more relevant to the task at hand & more distributed to
the classes.

See LeCun's commentary as well. Pawns occur more frequently than most of
the other pieces. The frequency probably needs to be normalized in the
training datasets.

Shashank G. Sawant, PhD
*(352) 226-1688*
*sgsawant@gmail.com <sgsawant@gmail.com>*
linkedin.com/in/sgsawant

**Comment by aaparab** [comment]
- The link is now updated. 
- The idea of writing a small function to identify if a square is empty or not is very good, I can try it. 
- Another follow-up is to have a sub-architecture identify the color of the piece once the previous function identifies a nonempty square. 
- I would like to use as little computer vision techniques as possible. I would rather have a slightly complex CNN architecture do the above two functions. 
- The other idea of weighing pieces by frequency of occurrence is not so good for this problem because these are not valid positions in a chess game. They are random auto-generated positions. 
- The question I'm posing here is mainly of how to design (the last layer of) a neural network assuming a non-empty square is found.
